src/hamiltonian.jl

```diff
@@ -42,16 +42,19 @@ end
 ∂H∂r(h::Hamiltonian{<:UnitEuclideanMetric,<:GaussianKinetic}, r::AbstractVecOrMat) = copy(r)
 ∂H∂r(h::Hamiltonian{<:DiagEuclideanMetric,<:GaussianKinetic}, r::AbstractVecOrMat) =
```

Comment:
Also, does this not suffer from the same issue? Is ComponentArrays.jl "clever" enough to realize this is a diagonal matrix and so there's no "mixing" between the components?

Reply:
The diagonal matrix case does not have the problem, because of the broadcasted multiplication. The type of `r` is maintained in this case. The test in line 89 of /test/hamiltonian.jl should cover this case.